src/mono/mono/mini/interp/transform.c

```diff
@@ -2504,6 +2504,42 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
 		// and always return false for IsDynamicCodeCompiled
 		if (!strcmp (tm, "get_IsDynamicCodeCompiled"))
 			*op = MINT_LDC_I4_0;
+#if defined(TARGET_X86) || defined(TARGET_AMD64)
+	} else if (in_corlib &&
+			!strncmp ("System.Runtime.Intrinsics.X86", klass_name_space, 29) &&
+			(!strcmp (klass_name, "Lzcnt")
+#if defined(TARGET_AMD64)
+			|| (!strcmp (klass_name, "X64") &&
+			m_class_get_nested_in (target_method->klass) &&
+			!strcmp (m_class_get_name (m_class_get_nested_in (target_method->klass)), "Lzcnt"))
+#endif
+			)) {
+		if (!strcmp (tm, "get_IsSupported")) {
+			*op = MINT_LDC_I4_1;
```

Comment:
This shouldn't be constant one, rather it should be returning 0 or 1 based on the relevant cached CPUID check

Reply:
Maybe I'm misunderstanding. We have the default condition where `get_IsSupported` for these intrinsics is replaced with `0` here https://github.com/dotnet/runtime/pull/82827/files#diff-45aa9a391314f1584abf38f453616b57fb3981160da30f96a51f11727d0c5472L2513. This PR changes it to become 1 on some arches (the check is done before the fallback), but the rest of the platforms are not affected and they still hit the default not supported case.

Reply:
The issue is that you're not validating the hardware supports the ISA, you're just returning 1.

This violates the contract of the API and may lead to user code faulting or having other issues depending on exactly what they were doing.

All of the platform specific IsSupported apis, where support exists, needs to be returning the result based on what the underlying CPU actually reports.

Correspondingly, it needs to also ensure the transitive closure is correct as well. That is, since x86.Lzcnt inherits from X86.x86Base, it reporting IsSupported=true implies that X86Base.IsSupported as well.

X86.Lzcnt.X64.IsSupported implies the same for X86.Lzcnt and X86.X86Base.X64

Reply:
Fair enough. I've removed handling of that class. Seems like overkill to implement it. I initially implemented it only for support of Log2, but since I ended up also intrinsifying it, doesn't seem that useful anyway.